plugins/wistron-dock/fu-wistron-dock-device.c

```diff
@@ -0,0 +1,944 @@
+		guint8 device_cnt = 0x0;
+		if (!fu_memread_uint8_safe(buf,
+					   sizeof(buf),
+					   G_STRUCT_OFFSET(FuWistronDockWdit, device_cnt),
+					   &device_cnt,
+					   error))
+			return FALSE;
+		if (!fu_wistron_dock_device_parse_wdit_img(self,
+							   buf,
+							   sizeof(buf),
+							   MIN(device_cnt, 32),
+							   error))
+			return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_wistron_dock_device_setup(FuDevice *device, GError **error)
+{
+	FuWistronDockDevice *self = FU_WISTRON_DOCK_DEVICE(device);
+
+	/* FuUsbDevice->setup */
+	if (!FU_DEVICE_CLASS(fu_wistron_dock_device_parent_class)->setup(device, error))
+		return FALSE;
+
+	if (!fu_wistron_dock_device_ensure_mcuid(self, error)) {
+		g_prefix_error(error, "failed to get MCUID: ");
+		return FALSE;
+	}
+	if (!fu_wistron_dock_device_ensure_bbinfo(self, error)) {
+		g_prefix_error(error, "failed to get BBINFO: ");
+		return FALSE;
+	}
+	if (!fu_wistron_dock_device_ensure_userinfo(self, error)) {
+		g_prefix_error(error, "failed to get USERINFO: ");
+		return FALSE;
+	}
+	if (!fu_wistron_dock_device_ensure_wdit(self, error)) {
+		g_prefix_error(error, "failed to get USERINFO: ");
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_wistron_dock_device_detach(FuDevice *device, FuProgress *progress, GError **error)
+{
+	FuWistronDockDevice *self = FU_WISTRON_DOCK_DEVICE(device);
+	guint8 cmd[8] = {FU_WISTRON_DOCK_ID_DOCK_IMG_CONTROL, FU_WISTRON_DOCK_CMD_DFU_ENTER};
+
+	/* sanity check */
+	if (fu_device_has_flag(device, FWUPD_DEVICE_FLAG_IS_BOOTLOADER)) {
+		g_debug("already in bootloader mode, skipping");
+		return TRUE;
+	}
+	if (!fu_wistron_dock_device_control_write(self, cmd, sizeof(cmd), FALSE, error))
+		return FALSE;
+	return fu_wistron_dock_device_ensure_wdit(self, error);
+}
+
+static gboolean
+fu_wistron_dock_device_attach(FuDevice *device, FuProgress *progress, GError **error)
+{
+	FuWistronDockDevice *self = FU_WISTRON_DOCK_DEVICE(device);
+	guint8 cmd[8] = {FU_WISTRON_DOCK_ID_DOCK_IMG_CONTROL, FU_WISTRON_DOCK_CMD_DFU_EXIT};
+	g_autoptr(FwupdRequest) request = fwupd_request_new();
+
+	/* sanity check */
+	if (!fu_wistron_dock_device_ensure_wdit(self, error))
+		return FALSE;
+	if (!fu_device_has_flag(device, FWUPD_DEVICE_FLAG_IS_BOOTLOADER)) {
+		g_debug("already in runtime mode, skipping");
+		return TRUE;
+	}
+	if (!fu_wistron_dock_device_control_write(self, cmd, sizeof(cmd), FALSE, error))
+		return FALSE;
+
+	/* the user has to do something */
+	fwupd_request_set_kind(request, FWUPD_REQUEST_KIND_IMMEDIATE);
+	fwupd_request_set_id(request, FWUPD_REQUEST_ID_REMOVE_REPLUG);
+	fwupd_request_add_flag(request, FWUPD_REQUEST_FLAG_ALLOW_GENERIC_MESSAGE);
+	fwupd_request_set_message(request,
+				  "The update will continue when the device USB cable has been "
+				  "unplugged and then re-inserted.");
+	fu_device_emit_request(device, request);
+	fu_device_add_flag(device, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
```

Comment:
If this is the behavior it uses; then I think you should be adding the flag that it's usable while it updates.

Reply:
Indeed, good catch, fixed.